src/bin/ipa_bench/gen.rs

```diff
@@ -0,0 +1,428 @@
+struct TriggerEvent {
+    event: Event,
+    value: u32,
+    zkp: String,
+}
+
+enum EventType {
+    S(SourceEvent),
+    T(TriggerEvent),
+    ES(ESourceEvent),
+    ET(ETriggerEvent),
+}
+
+// TODO: Currently, users are mutually exclusive in each ad loop (i.e. User A in ad X will never appear in other ads).
+// We need to generate events from same users across ads (but how often should a user appear in different ads?)
+// "Ads" doesn't mean FB's L3 ads. It could be ads from different businesses.
+
+pub fn generate_events(
+    total_count: u32,
+    epoch: u8,
+    secret_share: bool,
+    seed: &Option<u64>,
+    out: &mut Box<dyn io::Write>,
+) -> (u32, u32) {
+    let mut rng = match seed {
+        None => ChaCha20Rng::from_entropy(),
+        Some(seed) => ChaCha20Rng::seed_from_u64(*seed),
+    };
+    debug!("seed: {:?}", rng.get_seed());
+
+    // Separate RNG for generating secret shares
+    let mut ss_rng = match seed {
+        None => ChaCha20Rng::from_entropy(),
+        Some(seed) => ChaCha20Rng::seed_from_u64(*seed),
+    };
+
+    let sample = Sample::new();
+
+    let mut ad_count = 0;
+    let mut event_count = 0;
+    let mut s_count = 0;
+    let mut t_count = 0;
+
+    // Simulate impressions and conversions from an ad.
+    // We define "ad" as a group of impressions and conversions from targeted users who are selected by predefined
+    // breakdowns such as age, gender and locations.
+    loop {
+        ad_count += 1;
+        debug!("ad: {}", ad_count);
+
+        // TODO: 99.97% queries in ads manager account for L1-3 breakdown only. For now, we'll do 1 ad = 1 breakdown key
+        let ad_id: u32 = rng.gen();
+
+        // Number of unique people who saw the ad
+        let reach = sample.reach_per_ad(&mut rng);
+        debug!("reach: {}", reach);
+
+        // CVR for the ad
+        let cvr = sample.cvr_per_ad_account(&mut rng);
+        debug!("CVR: {}", cvr);
+
+        for _ in 0..reach {
+            // # of devices == # of matchkeys
+            let devices = sample.devices_per_user(&mut rng);
+            trace!("devices per user: {}", devices);
+
+            let impressions = sample.impression_per_user(&mut rng);
+            trace!("impressions per user: {}", impressions);
+
+            // Probabilistically decide whether this user has converted or not
+            let conversions = if Bernoulli::new(cvr).unwrap().sample(&mut rng) {
+                sample.conversion_per_user(&mut rng)
+            } else {
+                0
+            };
+            trace!("conversions per user: {}", conversions);
+
+            let events = gen_events(
+                devices,
+                impressions,
+                conversions,
+                epoch,
+                &ad_id.to_string(),
+                secret_share,
+                &sample,
+                &mut rng,
+                &mut ss_rng,
+            );
+
+            for e in events {
+                let json_string = match e {
```

Comment:
I think serde supports untagged representation of enums if that is what you want here: https://serde.rs/enum-representations.html